kornia/feature/disk/detector.py

```diff
@@ -0,0 +1,150 @@
+
+    return proposals, accept_mask, logp
+
+
+def nms(signal: Tensor, window_size=5, cutoff=0.0) -> Tensor:
+    if window_size % 2 != 1:
+        raise ValueError(f'window_size has to be odd, got {window_size}')
+
+    _, ixs = F.max_pool2d(signal, kernel_size=window_size, stride=1, padding=window_size // 2, return_indices=True)
+
+    # FIXME UPSTREAM: a workaround wrong shape of `ixs` until
+    # https://github.com/pytorch/pytorch/issues/38986
+    # is fixed
+    if len(ixs.shape) == 4:
+        assert ixs.shape[0] == 1
+        ixs = ixs.squeeze(0)
+
+    h, w = signal.shape[1:]
+    coords = torch.arange(h * w, device=signal.device).reshape(1, h, w)
+    nms = ixs == coords
+
+    if cutoff is None:
+        return nms
+    else:
+        return nms & (signal > cutoff)
+
+
+class Detector:
+    def __init__(self, window=8):
+        self.window = window
+
+    def _tile(self, heatmap: Tensor) -> Tensor:
+        """Divides the heatmap `heatmap` into tiles of size (v, v) where v==self.window. The tiles are flattened,
+        resulting in the last.
+
+        dimension of the output T == v * v.
+        """
+        v = self.window
+        b, c, h, w = heatmap.shape
+
+        assert heatmap.shape[2] % v == 0
+        assert heatmap.shape[3] % v == 0
+
+        return heatmap.unfold(2, v, v).unfold(3, v, v).reshape(b, c, h // v, w // v, v * v)
+
+    def sample(self, heatmap: Tensor) -> np.ndarray:
+        """Implements the training-time grid-based sampling protocol."""
+        v = self.window
+        dev = heatmap.device
+        B, _, H, W = heatmap.shape
+
+        assert H % v == 0
+        assert W % v == 0
+
+        # tile the heatmap into [window x window] tiles and pass it to
+        # the categorical distribution.
+        heatmap_tiled = self._tile(heatmap).squeeze(1)
+        proposals, accept_mask, logp = point_distribution(heatmap_tiled)
+
+        # create a grid of xy coordinates and tile it as well
+        cgrid = torch.stack(
+            torch.meshgrid(torch.arange(H, device=dev), torch.arange(W, device=dev), indexing='ij')[::-1], dim=0
+        ).unsqueeze(0)
+
+        # extract xy coordinates from cgrid according to indices sampled
+        # before
+        xys = select_on_last(
+            self._tile(cgrid).repeat(B, 1, 1, 1, 1),
+            # unsqueeze and repeat on the (xy) dimension to grab
+            # both components from the grid
+            proposals.unsqueeze(1).repeat(1, 2, 1, 1),
+        ).permute(
+            0, 2, 3, 1
+        )  # -> bhw2
+
+        keypoints = []
+        for i in range(B):
+            mask = accept_mask[i]
+            keypoints.append(Keypoints(xys[i][mask], logp[i][mask]))
+
+        return np.array(keypoints, dtype=object)
+
+    def nms(self, heatmap: Tensor, n: Optional[int] = None, **kwargs) -> np.ndarray:
+        """Inference-time nms-based detection protocol."""
+        heatmap = heatmap.squeeze(1)
+        nmsed = nms(heatmap, **kwargs)
```

Comment:
i don't see a reason why this cannot be just `list[Keypoint]` since what you only do is random access to the list
+ suggest to migrate typing to `from __future__ import annotate`